File: plugins/plugin-scratchpad/python/elizaos_plugin_scratchpad/actions/list.py
"""SCRATCHPAD_LIST action - List all scratchpad entries."""

from __future__ import annotations

import logging
from dataclasses import dataclass
from typing import TYPE_CHECKING

from elizaos_plugin_scratchpad.service import create_scratchpad_service

if TYPE_CHECKING:
    from elizaos.types import HandlerCallback, HandlerOptions, IAgentRuntime, Memory, State

logger = logging.getLogger(__name__)


@dataclass
class ListResult:
    """Result of the SCRATCHPAD_LIST action."""

    success: bool
    text: str
    error: str | None = None


async def handle_scratchpad_list(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
    options: HandlerOptions | None = None,
    callback: HandlerCallback | None = None,
    responses: list[Memory] | None = None,
) -> ListResult | None:
    """Handle the SCRATCHPAD_LIST action."""
    try:
        service = create_scratchpad_service(runtime)
        entries = await service.list()

        if not entries:
            msg = "You don't have any scratchpad entries yet. Use SCRATCHPAD_WRITE to create one."
            if callback:
                await callback(
                    {
                        "text": msg,
                        "actions": ["SCRATCHPAD_LIST_EMPTY"],
                        "source": message.content.source if message.content else None,
                    }
                )
            return ListResult(success=True, text="No entries")

        lines: list[str] = []
        for i, entry in enumerate(entries):
            tags_str = f" [{', '.join(entry.tags)}]" if entry.tags else ""
            lines.append(
                f"{i + 1}. **{entry.title}** ({entry.id}){tags_str}\n"
                f"   _Modified: {entry.modified_at.strftime('%Y-%m-%d')}_"
            )

        list_text = "\n".join(lines)
        success_msg = f"**Your Scratchpad Entries** ({len(entries)} total):\n\n{list_text}"

        if callback:
            await callback(
                {
                    "text": success_msg,
                    "actions": ["SCRATCHPAD_LIST_SUCCESS"],
                    "source": message.content.source if message.content else None,
                }
            )

        return ListResult(success=True, text=success_msg)

    except Exception as exc:
        error_msg = f"Failed to list scratchpad entries: {exc}"
        logger.error("[ScratchpadList] Error: %s", exc)
        if callback:
            await callback(
                {
                    "text": error_msg,
                    "actions": ["SCRATCHPAD_LIST_FAILED"],
                    "source": message.content.source if message.content else None,
                }
            )
        return ListResult(success=False, text="Failed to list scratchpad entries", error=str(exc))


async def validate_scratchpad_list(
    runtime: IAgentRuntime,
    message: Memory,
    state: State | None = None,
) -> bool:
    """Always validate - listing is always allowed."""
    return True


SCRATCHPAD_LIST_ACTION: dict[str, object] = {
    "name": "SCRATCHPAD_LIST",
    "similes": [
        "LIST_NOTES",
        "SHOW_NOTES",
        "MY_NOTES",
        "ALL_NOTES",
        "SHOW_SCRATCHPAD",
    ],
    "description": "List all scratchpad entries with their titles, IDs, tags, and modification dates.",
    "examples": [
        [
            {
                "name": "{{name1}}",
                "content": {"text": "Show me all my saved notes"},
            },
            {
                "name": "{{name2}}",
                "content": {
                    "text": "**Your Scratchpad Entries** (2 total):\n\n1. **Meeting Notes** (meeting-notes)\n2. **Project Ideas** (project-ideas)",
                    "actions": ["SCRATCHPAD_LIST_SUCCESS"],
                },
            },
        ],
    ],
}

File: packages/core/src/features/advanced-capabilities/clipboard/actions/list.ts
import {
	type Action,
	type HandlerCallback,
	type HandlerOptions,
	type IAgentRuntime,
	logger,
	type Memory,
	type State,
} from "../../../../types/index.ts";
import { hasActionContextOrKeyword } from "../../../../utils/action-validation.ts";
import { createClipboardService } from "../services/clipboardService.ts";
import { requireActionSpec } from "../specs.ts";

const spec = requireActionSpec("CLIPBOARD_LIST");

export const clipboardListAction: Action = {
	name: spec.name,
	contexts: ["files", "knowledge", "agent_internal"],
	roleGate: { minRole: "ADMIN" },
	similes: spec.similes ? [...spec.similes] : [],
	description: spec.description,
	parameters: [],

	validate: async (
		_runtime: IAgentRuntime,
		message: Memory,
		state?: State,
		_options?: HandlerOptions,
	): Promise<boolean> => {
		return hasActionContextOrKeyword(message, state, {
			contexts: ["files", "knowledge", "agent_internal"],
			keywords: ["clipboard", "list notes", "show notes", "saved notes"],
		});
	},

	handler: async (
		runtime: IAgentRuntime,
		message: Memory,
		_stateFromTrigger: State | undefined,
		_options: HandlerOptions | undefined,
		callback?: HandlerCallback,
		_responses?: Memory[],
	) => {
		try {
			const service = createClipboardService(runtime);
			const entries = await service.list();

			if (entries.length === 0) {
				if (callback) {
					await callback({
						text: "You don't have any clipboard entries yet. Use CLIPBOARD_WRITE to create one.",
						actions: ["CLIPBOARD_LIST_EMPTY"],
						source: message.content.source,
					});
				}
				return { success: true, text: "No entries", entries: [] };
			}

			const listText = entries
				.map((e, i) => {
					const tagsStr = e.tags?.length ? ` [${e.tags.join(", ")}]` : "";
					return `${i + 1}. **${e.title}** (${e.id})${tagsStr}\n   _Modified: ${e.modifiedAt.toLocaleDateString()}_`;
				})
				.join("\n");

			const successMessage = `**Your Clipboard Entries** (${entries.length} total):\n\n${listText}`;

			if (callback) {
				await callback({
					text: successMessage,
					actions: ["CLIPBOARD_LIST_SUCCESS"],
					source: message.content.source,
				});
			}

			return { success: true, text: successMessage, entries };
		} catch (error) {
			const errorMsg = error instanceof Error ? error.message : String(error);
			logger.error("[ClipboardList] Error:", errorMsg);
			if (callback) {
				await callback({
					text: `Failed to list clipboard entries: ${errorMsg}`,
					actions: ["CLIPBOARD_LIST_FAILED"],
					source: message.content.source,
				});
			}
			return { success: false, text: "Failed to list clipboard entries" };
		}
	},

	examples: [],
};

export default clipboardListAction;
